### DISCUSSION TITLE
Getting TypeError: Cannot read properties of null (reading 'useState')

### DISCUSSION THREAD

**Discussion by rahulsivalenka** [originjs/vite-plugin-federation#221] [Q&A]
I am trying to setup two **_React + Vite + Typescript_** applications, one host, the other remote using the vite-plugin-federation. After building and running both applications using `vite preview`, when running the host app I am getting the following error:

![image](https://user-images.githubusercontent.com/15724285/180715909-5f01752d-354a-42a4-971e-dbc0f21f3b1e.png)

My Vite config for each of the apps are as follows:

# Host app
```js
import { defineConfig } from "vite";
import react from "@vitejs/plugin-react";
import federation from "@originjs/vite-plugin-federation";

// HOST
// https://vitejs.dev/config/
export default defineConfig({
  preview: {
    port: 3000,
  },
  build: {
    target: "esnext",
    minify: false,
  },
  plugins: [
    react(),
    federation({
      remotes: {
        remote: "http://localhost:3001/assets/remoteEntry.js",
      },
      shared: {
        react: {
          requiredVersion: "^18.2.0",
          singleton: true,
          // version: "18.2.0",
        },
        "react-dom": {
          requiredVersion: "^18.2.0",
          singleton: true,
        },
      },
    }),
  ],
});
```

# Remote app
```js
import { defineConfig } from "vite";
import react from "@vitejs/plugin-react";
import federation from "@originjs/vite-plugin-federation";

// REMOTE
// https://vitejs.dev/config/
export default defineConfig({
  preview: {
    port: 3001,
  },
  build: {
    target: "esnext",
    minify: false,
  },
  plugins: [
    react(),
    federation({
      filename: "remoteEntry.js",
      exposes: {
        "./App": "./src/App.tsx",
      },
      shared: {
        react: {
          requiredVersion: "^18.2.0",
          singleton: true,
        },
        "react-dom": {
          requiredVersion: "^18.2.0",
          singleton: true,
        },
      },
    }),
  ],
});
```

I have tried various combinations of the config but nothing works by referring to [React ESM example](github.com/originjs/vite-plugin-federation/tree/main/packages/examples/simple-react-esm) (which uses rollup), also looked at [vue3-demo-webpack-esm-esm](https://github.com/originjs/vite-plugin-federation/tree/main/packages/examples/vue3-demo-webpack-esm-esm) example for reference. Next step, once this works, is for me to try to make **_React + Webpack + Typescript_** app as host and _**React + Vite + Typescript**_ app as remote.

Plz help me understand what I might be missing. 

Find repro [here](https://github.com/rahulsivalenka/vite-module-federation-test). I have committed the `dist` folders of the apps here for reference of the files generated after the build.

**Comment by rahulsivalenka** [comment]
Hello. Any suggestions that could help solve this problem?

**Reply by Jiannan-dev** [reply]
I think this issue #173 will help you

**Reply by rahulsivalenka** [reply]
Thank you for your reply.

Yes, I found that it works after building and serving them with `vite preview`. [Here's my trial repo](https://github.com/rahulsivalenka/module-federation-test) that worked for me with both host and remote using ViteJS.

I moved on to setting up a host using webpack and using the same remote as above (see [branch:webpack-vite-trial](https://github.com/rahulsivalenka/module-federation-test/tree/webpack-vite-trial)). However, in this case I am getting the exact error as #173.

I tried both the suggestions suggested in the issue 173's comments:

1. Using `React.useState` instead of destructuring `useState` from react.
2. Building both webpack host and vite remote app and serving them.

In both cases I am getting the same issue `TypeError: Cannot read properties of undefined (reading: 'exports')`.
![image](https://user-images.githubusercontent.com/15724285/182831018-bb1b51cc-f5b5-4d2e-85a0-bfbf45a28ded.png)

Vite as a host is giving the right imported react module structure which seems to be compatible with what is expected in the `__federation_expose_App_....js` file.
![image](https://user-images.githubusercontent.com/15724285/182832721-e5394642-558b-43e2-8195-79b8c028dcad.png)

When using webpack as host, the structure of the react module shared by webpack is different and this is causing the above TypeError issue.
![image](https://user-images.githubusercontent.com/15724285/182833231-d7a694e1-9bfd-46f0-afc7-11e40f6a7445.png)


I have configured webpack to output as module:
```js
// webpack.config.js
//...
return {
  target: 'es2020',
  experiments: {
    outputModule: true,
  },
  plugins: [
    new ModuleFederationPlugin({
      name: 'host-webpack',
      library: { type: 'module' },
      filename: 'remoteEntry.js',
      remotes: {
        // this is the URL of the vite remote app running in preview mode
        remote: 'http://localhost:8082/assets/remoteEntry.js',
      },
      shared: {
        ...pkg.dependencies,
        react: {
          // eager: true,
          // singleton: true,
          requiredVersion: pkg.dependencies['react'],
        },
        'react-dom': {
          // eager: true,
          // singleton: true,
          requiredVersion: pkg.dependencies['react-dom'],
        },
      },
    }),
    //...
  ],
  //...
};
//...
```
(see full config [here](https://github.com/rahulsivalenka/module-federation-test/blob/webpack-vite-trial/host-webpack/config/webpack.config.js#L546) which was generated by ejecting a new CRA app)

Is it something in my webpack configuration  that is causing the structure of the imported React module to change? 

Again, thank you so much for taking the time to help me out.

**Reply by Jiannan-dev** [reply]
It looks like the webpack generated chunk and vite (esbuild and rollup to be exact) generated chunk is not consistent, later I clone your repo to see the specific problem, thank you for your feedback

**Reply by rahulsivalenka** [reply]
Looking forward to a newer version. Thanks for your help.

**Reply by rlesniak** [reply]
Any news here? I am facing identical issue... @ruleeeer

**Comment by chitralverma** [comment]
Any help on this ?

**Reply by chitralverma** [reply]
@flyfishzy @ruleeeer any help?

**Comment by Arpan1089** [comment]
Any work around to fix this ??

**Comment by robertovg** [comment]
Any updates on this? I'm facing the same problem, components, if remote components exported are not using React... which is un-useful.

**Comment by robertovg** [comment]
I have the solution for this: https://github.com/originjs/vite-plugin-federation/issues/242#issuecomment-1256430601

We should have a working solution compatible with the last version of the library though.
